Provis.UnitTests/Core/Services/TaskServiceTests.cs

```diff
@@ -0,0 +1,1234 @@
+            var taskAssign = AssignOnTaskDTO;
+
+            SetupJoinTask(taskMock, isMember, assigned);
+
+            Func<Task> act = () => _taskService.JoinTaskAsync(taskAssign, userId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.UserAlreadyAssigned);
+        }
+
+        [Test]
+        [TestCase("2", true, false)]
+        public async Task JoinTask_AssignUser_ReturnTaskComplete(string userId, bool isMember, bool assigned)
+        {
+            var taskMock = GetWorkspaceTask;
+            var taskAssignDTO = AssignOnTaskDTO;
+            var taskAssign = new UserTask()
+            {
+                TaskId = taskAssignDTO.Id,
+                IsUserDeleted = false,
+                Task = taskMock,
+                UserId = taskAssignDTO.AssignedUser.UserId,
+                UserRoleTagId = taskAssignDTO.AssignedUser.RoleTagId
+            };
+
+            SetupJoinTask(taskMock, isMember, assigned);
+            SetupUserTaskSaveChangesAsync();
+            SetupUserTaskAddAsync(taskAssign);
+            SetupMetricsIncrement();
+            _mapperMock.SetupMap(taskAssignDTO, taskAssign);
+
+            var result = _taskService.JoinTaskAsync(taskAssignDTO, userId);
+
+            result.IsCompleted.Should().BeTrue();
+            result.IsCompletedSuccessfully.Should().BeTrue();
+
+            await Task.CompletedTask;
+        }
+
+        [Test]
+        [TestCaseSource("ChangeStatus")]
+        public async Task ChangeStatus_TaskNotFound_ThrowHttpException(string userId, TaskChangeStatusDTO changeDTO)
+        {
+            WorkspaceTask taskMock = null;
+            var taskAssign = AssignOnTaskDTO;
+
+            SetupTaskGetByKeyAsync(taskMock);
+            Func<Task> act = () => _taskService.ChangeTaskStatusAsync(changeDTO, userId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.NotFound)
+                .WithMessage(ErrorMessages.TaskNotFound);
+        }
+
+        [Test]
+        [TestCaseSource("ChangeStatus")]
+        public async Task ChangeStatus_WithMailing_ReturnTaskComplete(string userId, TaskChangeStatusDTO changeDTO)
+        {
+            var taskMock = GetWorkspaceTask;
+            var userMock = GetUser;
+            var workspaceMock = GetWorkspace;
+            var statusHistoryMock = StatusHistories[1];
+
+            SetupChangeInfo(userId, userMock, taskMock, Emails, workspaceMock, GetClientUrl);
+            SetupChangeStatusSendManyMailsASync();
+            SetupMetricsIncrement();
+            SetupMetricsDecrement();
+            SetupHistoryAddAsync(statusHistoryMock);
+
+            var result = Task.Run(() =>
+                _taskService.ChangeTaskStatusAsync(changeDTO, userId));
+            result.Wait();
```

Comment:
Remove this necessary!

Reply:
I don't know how to change this. Because of Task, my test method ends earlier and return an incorrect result.